### PR TITLE
build: make Therpists search

### PR DESCRIPTION
**Pull request by raghdarizq**
- this Route should fetch therapists
-  getAllTherapist: Search for a therapists by name and cheek if his active 
The code allows for searching therapists by name and implementing pagination to retrieve a specific number of therapists per page. If no name is provided in the query, it retrieves all therapists without any filtering.

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,12 +1,13 @@
...
 import { clientError, serverError } from './middlewares';
+import therapists from './routes';
 
 const app = express();
 
 app.get('/', (req: Request, res: Response) => {
   res.send('Hello World!');
 });
-
+app.use(therapists);
```
Comment (LinaYahya) [inline src/app.ts:10]: Here we need to use our main Router, at `/router/index.ts` please import all subrouter and use it there

File: src/routes/therapists.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import express from 'express';
+import getAllTherapists from '../controllers';
+
+const therapists = express.Router();
+
+therapists.get('/therapists', getAllTherapists);
+export default therapists;
+// http://localhost:8080/therapists?q=Al&page=0
```
Comment (LinaYahya) [inline src/routes/therapists.ts:8]:
```suggestion
```

File: src/services/therapists.ts [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        attributes: ['fullName', 'isActive'],
+        where: {
+          isActive: true,
+          fullName: {
+            [Op.like]: `%${name}%`,
+          },
+        },
+      },
+    ],
+    attributes: ['profileImg', 'major', 'hourlyRate'],
```
Comment (LinaYahya) [inline src/services/therapists.ts:19]: we need to retrieve therapistID so we can redirect user to therapist profile

File: src/services/therapists.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { Op } from 'sequelize';
+import { Therapist, User } from '../models';
+
+const getAllTherapist = async (name: string, page: number, limit: number) => {
```
Comment (LinaYahya) [inline src/services/therapists.ts:4]: we don't neet to have limit here as a parameter, Insted let us import `Limit` and use it Here

File: src/controllers/therapists.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    const pageNumber = parseInt(page as string, 2) || 1;
+    const limit = 8;
+
+    const therapists = await getAllTherapist(q as string || ' ', pageNumber, limit);
+
+    if (!therapists.length) {
+      res.status(204).json({});
+    }
+
+    res.status(200).json(therapists);
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:16]:
```suggestion
    res.json(therapists);
```
The default status code going to be `200` So there is no need to specify it, as well as response payload doesn't follow what we agreed on

File: src/controllers/therapists.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+const getAllTherapists = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { q, page } = req.query;
+    const pageNumber = parseInt(page as string, 2) || 1;
+    const limit = 8;
+
+    const therapists = await getAllTherapist(q as string || ' ', pageNumber, limit);
+
+    if (!therapists.length) {
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:12]: what if we have zero length therapist, What do u expect to happen,
Can u please trace the whole flow

File: src/controllers/therapists.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
+import { NextFunction, Request, Response } from 'express';
+import getAllTherapist from '../services';
+
+const getAllTherapists = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { q, page } = req.query;
+    const pageNumber = parseInt(page as string, 2) || 1;
+    const limit = 8;
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:8]: let us declare this at separate file at config folder, and please UPPERCASE it as this is a constant ,
```js
const THERAPISTS_LIMIT = 8;
```

File: src/controllers/therapists.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    const therapists = await getAllTherapist(q as string || ' ', pageNumber, limit);
+
+    if (!therapists.length) {
+      res.status(204).json({});
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:13]:
```suggestion
      res.status(204).json({data: [], message: 'No therapist found'});
```

File: src/controllers/therapists.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import { NextFunction, Request, Response } from 'express';
+import getAllTherapist from '../services';
+
+const getAllTherapists = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { q, page } = req.query;
+    const pageNumber = parseInt(page as string, 2) || 1;
```
[the remark this answers is not in the document]
Comment (LinaYahya) [reply src/controllers/therapists.ts:7]: can you please check having page as a string?
Can u please try to print pageNumber when page is `10` or `zero`

**Review by MohannadSabaa — APPROVED** [review]

**Review by LinaYahya — CHANGES_REQUESTED** [review]
please check [this PR](https://github.com/GSG-G13/ntherapy-express/pull/12
) comments as well,
Also we have to start with offest `0`

File: __tests__/therapists.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import request from 'supertest';
+import app from '.././src/app';
+import buildDB from "../src/db/build";
```
Comment (ahmedmurtaja) [inline __tests__/therapists.ts:3]: use SingleQoutes

File: __tests__/therapists.ts [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+import dbConnection from "../src/db/connection";
```
Comment (ahmedmurtaja) [inline __tests__/therapists.ts:4]: use SingleQoutes

**Review by YasserAmro0 — APPROVED** [review]

File: src/controllers/therapists.ts [hunk 12]
```diff
@@ -1,27 +1,48 @@
...
+const getAllTherapists = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { q, page } = req.query;
+    const pageNumber = Number(page) || 1;
+
+    const therapists = await getAllTherapist(q as string || ' ', pageNumber, THERAPISTS_LIMIT);
+
+    if (therapists.length === 0) {
+      return res.status(204).json({ message: 'No therapists found' });
+    }
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:37]: I guess no need to check therapists' length, SO if no therapists are found we'll end with empty array as a result, and no problem with that, while here we can't handle response as status `204` responded with no content, So I'll not be able to catch response result

File: src/controllers/therapists.ts [hunk 13]
```diff
@@ -1,27 +1,48 @@
...
+    next(error);
+  }
+};
+
+const getAllTherapists = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { q, page } = req.query;
+    const pageNumber = Number(page) || 1;
+
+    const therapists = await getAllTherapist(q as string || ' ', pageNumber, THERAPISTS_LIMIT);
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:33]:
```suggestion
    const therapists = await getAllTherapist(q as string || '', pageNumber, THERAPISTS_LIMIT);
```
Why to have default value as a space, replace it please with empty string

File: src/services/therapists.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+        where: {
+          isActive: true,
+          fullName: {
+            [Op.like]: `%${name}%`,
+          },
+        },
+      },
+    ],
+    attributes: ['profileImg', 'major', 'hourlyRate', 'userId'],
+    limit,
```
Comment (LinaYahya) [inline src/services/therapists.ts:34]:
```suggestion
    limit: THERAPISTS_LIMIT,
```

File: src/db/connection.ts [hunk 15]
```diff
@@ -16,4 +16,6 @@ const sequelize = new Sequelize(DB_URL, {
   logging: false,
 });
 
+export const THERAPISTS_LIMIT = 8;
```
Comment (LinaYahya) [inline src/db/connection.ts:19]: please move this to config/constants.ts and import from there

File: src/controllers/therapists.ts (hunk 13 above)
Comment (LinaYahya) [inline src/controllers/therapists.ts:33]:
```suggestion
    const therapists = await getAllTherapist(q as string || ' ', pageNumber);
```

File: src/controllers/therapists.ts [hunk 16]
```diff
@@ -1,27 +1,48 @@
...
+    }
+  } catch (error) {
+    next(error);
+  }
+};
+
+const getAllTherapists = async (req: Request, res: Response, next: NextFunction) => {
+  try {
+    const { q, page } = req.query;
+    const pageNumber = Number(page) || 1;
```
Comment (LinaYahya) [inline src/controllers/therapists.ts:31]: please check if the page is a number or not, If not responded with 400 bad request

**Review by LinaYahya — CHANGES_REQUESTED** [review]

**Review by LinaYahya — APPROVED** [review]

**Review by ahmedmurtaja — APPROVED** [review]

**Pull request merged**